### PR TITLE
samples: Added the Train until good pipeline

### PR DESCRIPTION
**Pull request by Ark-kun**
This sample demonstrates continuous training using a train-eval-check recursive loop.
The main pipeline trains the initial model and then gradually trains the model some more until the model evaluation metrics are good enough.

image

### CODE REVIEW

File: samples/core/train_until_good/train_until_good.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# This sample demonstrates continuous training using a train-eval-check recursive loop.
```
Comment (numerology) [inline samples/core/train_until_good/train_until_good.py:1]: Shall we add copyright statement at the top?

File: samples/core/train_until_good/train_until_good.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    metrics_task = calculate_regression_metrics_from_csv_op(
+        true_values=true_values,
+        predicted_values=predictions,
+    )
+
+    # Checking the metrics
+    with kfp.dsl.Condition(metrics_task.outputs['mean_squared_error'] > 0.01):
+        # Training some more
+        train_until_low_error(
+            starting_model=model,
```
Comment (numerology) [inline samples/core/train_until_good/train_until_good.py:47]: qq: I see you're using warm starting here. Is part of the reasons that without this caching will step in and mess up the recursion?

File: samples/core/train_until_good/train_until_good.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            training_data=training_data,
+            true_values=true_values,
+        )
+
+
+# The main pipleine trains the initial model and then gradually trains the model some more until the model evaluation metrics are good enough.
```
Comment (numerology) [inline samples/core/train_until_good/train_until_good.py:53]: nit: line too long

File: samples/core/train_until_good/train_until_good.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    training_data = chicago_taxi_dataset_op(
+        where='trip_start_timestamp >= "2019-01-01" AND trip_start_timestamp < "2019-02-01"',
+        select='tips,trip_seconds,trip_miles,pickup_community_area,dropoff_community_area,fare,tolls,extras,trip_total',
+        limit=10000,
+    ).output
+
+    # Preparing the true values
+    true_values_table = pandas_transform_csv_op(
+        table=training_data,
+        transform_code='''df = df[["tips"]]''',
```
Comment (numerology) [inline samples/core/train_until_good/train_until_good.py:65]: Do we need triple quotes here?

File: samples/core/train_until_good/train_until_good.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    train_until_low_error(
+        starting_model=first_model,
+        training_data=training_data,
+        true_values=true_values,
+    )
+
+
+if __name__ == '__main__':
+    kfp_endpoint=None
+    kfp.Client(host=kfp_endpoint).create_run_from_pipeline_func(train_until_good_pipeline, arguments={})
```
Comment (numerology) [inline samples/core/train_until_good/train_until_good.py:88]: Currently other core samples output a compiled pipeline yaml spec, which is more compatible with the current test infra. Shall we also make this sample output a yaml instead of kicking off an execution? So that it's easier to add sample test later on.

File: samples/core/train_until_good/train_until_good.py (hunk 2 above)
Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:47]: >Is part of the reasons that without this caching will step in and mess up the recursion?

I do not think so. The caching works great here and it does not know anything about recursion. Caching works on the lower execution level and is not affected by recursion, loops or conditionals.

I'm not sure I fully understand your question. Can you please help me understand it.

The first model training does not use warm start. Then the loop body continues training that model recursively (which is the goal of the sample).

Without continuing from the previous model, the model won't improve, right? If we abandoned the previous model, then why did we train it?

Not every continuous training pipeline will use this pattern. There can be a pipeline that compares production model, a cleanly-trained model and a mixed model and chooses the best one.

File: samples/core/train_until_good/train_until_good.py (hunk 4 above)
Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:65]: I can remove them. Triple quotes may be needed in the transform code since there are often quoted strings and also multi-line code.

File: samples/core/train_until_good/train_until_good.py (hunk 5 above)
Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:88]: Hmm. I'll think of the best way to do this. I wonder whether we can submit those kind of samples directly.

File: samples/core/train_until_good/train_until_good.py (hunk 2 above)
Comment (numerology) [reply samples/core/train_until_good/train_until_good.py:47]: > Without continuing from the previous model, the model won't improve, right?

Totally agree. Let me elaborate on my question (of course below is just an edge case, I am not suggesting users should do this):

If we hardcode the model uri inside the `train_until_low_error` graph component, and remove that from the function signature of `train_until_low_error`, and we make `xgboost_train_on_csv_op` pick up a model from the hard-coded uri and override that after finish. Then, each time this recursion will see exactly the same component input spec, so, by default the cache is hit and the actual execution will be skipped.

Did I miss something here?

Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:47]: > If we hardcode the model uri inside the `train_until_low_error` graph component, and remove that from the function signature of `train_until_low_error`, and we make `xgboost_train_on_csv_op` pick up a model from the hard-coded uri and override that after finish. Then, each time this recursion will see exactly the same component input spec, so, by default the cache is hit and the actual execution will be skipped.

I think this is true. In that case, each "iteration" is exactly the same. Normally, in recursion, the next iteration depends on previous (and as such is different from the previous iteration). But if we remove that data dependency, then all iterations become the same and so the caching system will skip them.

>pick up a model from the hard-coded uri and override that after finish.
This is important part I missed during the first reading.
Yes, the caching system cannot automatically detect that the data behind t he URI has changed. Doing this is not practical. The caching system is only guaranteed to work correctly when the data is immutable and the components are pure.

If the data is expected to change, then the task that downloads the model from the URI should be marked as volatile (`max_cache_staleness = "P0D"`), so that the cache is disabled.

An even better solution is to not overwrite the data behind the URIs. When using the usual data passing methods this cannot happen. Every new model should be written to a separate URI and this new URI should be passed to the next iteration.

Can you please describe a bigger scenario?
I think you are thinking about resolver-type scenarios.
They can be implemented and can even use cache selectively: The code should not overwrite the model itself, but can overwrite a blob *pointing* to that URI. For example `gs://.../models/prod/latest` has content `gs://.../models/prod/a4bf...7ba1`. Then the model downloading can be split into 1) Getting the active prod model URI (volatile) 2) Downloading from that URI. This way, if the model URI has not really changed, the cache would be used successfully. I'm finishing another sample pipeline demonstrating this workflow.

File: samples/core/train_until_good/train_until_good.py (hunk 5 above)
Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:88]: I think I'll solve this issue when I'll be adding this sample to sample tests.
I'm trying to keep the samples for users first and tests second.

Comment (numerology) [reply samples/core/train_until_good/train_until_good.py:88]: Sounds good.

File: samples/core/train_until_good/train_until_good.py (hunk 2 above)
Comment (numerology) [reply samples/core/train_until_good/train_until_good.py:47]: > Can you please describe a bigger scenario?

I think the bigger picture is not quite related to this CL, but in general how caching and even orchestration should be implemented, so please view this comment as non-blocking.

As you can see, caching behavior that is based on input/output spec, or component command line args is sensitive to some corner cases where component behavior has some **hidden** dependencies, and we need to rely on users doing the right thing to make sure things work out properly. For caching, if the artifact passed in can be associated with a fingerprint (which is a sort of MLMD property) then I think perhaps we can get rid of this.

Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:47]: >I think the bigger picture is not quite related to this CL, but in general how caching and even orchestration should be implemented, so please view this comment as non-blocking.

Thank you. I think these discussions are really insightful.

>caching behavior that is based on input/output spec, or component command line args is sensitive to some corner cases where component behavior has some hidden dependencies

This is true. But I'm not sure this problem can be solved.
You cannot be 100% sure what the user code will do unless you actually run that code.
The component can have non-deterministic behavior, random behavior (bagging, stochastic gradient descent) and can access arbitrary external data (E.g. today's news.). In general it's not possible to predict or prevent this non-deterministic behavior. The best that we could do is to make the containers hermetic and prevent them from accessing any external resources (services, internet, etc). But that would severely limit the usefulness.

>For caching, if the artifact passed in can be associated with a fingerprint (which is a sort of MLMD property) then I think perhaps we can get rid of this.

Here is a subtle, but important aspect: Unlike TFX's artifacts, KFP's intermediate artifacts are isolated and immutable. So the caching guarantees are significantly easier to hold. The components never write to the artifact repository directly and it usually cannot be easily accessed. It's not possible to pass an artifact with custom URI to the KFP component. The only way to add a new artifact is to execute a component which will produce the artifact (which will be stored at the unique immutable location). This seem to make the described problem impossible since the artifact data cannot really change. Without the possibility of data change, there is not much difference between fingerprint and unique id.

Now let's consider a slightly different problem: Caching and components that access external data that can change.
In general, it's not possible to determine the fingerprint of external data and we cannot force the fingerprint to be calculated by every piece of user code getting anything from the internet.
In KFP, the caching behavior is pretty easy: If a downstream component takes an artifact from an upstream component, then the downstream component will always execute if the upstream has executed. The caching is essentially based on the artifact and execution lineage. So, the user either forces the upstream downloader component to re-execute even when all arguments are the same (which in turn re-executed all downstream components) or the user does not force the upstream to re-execute and the cache is used when possible. There user is in control.

There is also a middle option which allows the pipeline author to perform the data check, but still use cache if the data has not changed. I've described a way to make it possible at the end of the previous comment.

So:
* For people with well-behaved pure components the caching just works.
* For people with volatile components, there is a simple way to disable caching for those components.
* There is also a slightly more complicated way to regain some caching support even for the volatile components.

Some action items for myself based on this discussion:

* Add a component that peeks the MD5 checksum of a GCS blob without downloading it
* Add a component that calculates a the data pointed by an HTTP URL
* Add a component that downloads the data from HTTP which optionally accepts the expected data hash. The downloaded data will be checked against the hash.
* Add a component that downloads the data from GCS which optionally accepts the expected data hash. The downloaded data will be checked against the hash.

With this components it would be pretty easy to implement data "peeking" in a cache-preserving way.
First the data hash is calculated. If the hash is the same as before, then the cache will be used downstream. Otherwise, the cache won't be used and the data will be downloaded and processed.

File: samples/core/train_until_good/train_until_good.py (hunk 1 above)
Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:1]: Added the license header.

File: samples/core/train_until_good/train_until_good.py (hunk 4 above)
Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:65]: Removed.

File: samples/core/train_until_good/train_until_good.py (hunk 3 above)
Comment (Ark-kun) [reply samples/core/train_until_good/train_until_good.py:69]: Done.

**Comment by Ark-kun** [comment]
I think I've resolved the comments and the tests now succeed.

**Comment by Ark-kun** [comment]
Friendly ping
